### DISCUSSION TITLE
Error on uploading send with custom expiration duration

### DISCUSSION THREAD

**Discussion by Strentz-Paul** [dani-garcia/vaultwarden#2024] [Q&A]
### Subject of the issue
Error on uploading send with custom expiration duration
### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.22.2
* Web-vault version: v2.21.1
* Running within Docker: true
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: false (X-Forwarded-For)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: PostgreSQL
* Database version: PostgreSQL 13.1 (Debian 13.1-1.pgdg100+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 8.3.0-6) 8.3.0, 64-bit
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** DOMAIN, SIGNUPS_ALLOWED, SIGNUPS_VERIFY, ADMIN_TOKEN

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_ip_header_enabled": true,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_max_conns": 10,
  "database_url": "**********://*********:**********************@********/",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://**.********.**",
  "domain_origin": "*****://**.********.**",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "enable_db_wal": true,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "invitation_org_name": "***",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "org_attachment_limit": null,
  "org_creation_users": "",
  "password_iterations": 100000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": true,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_explicit_tls": false,
  "smtp_from": "**@********.**",
  "smtp_from_name": "Bitwarden - ****",
  "smtp_host": "****.*****.**",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_ssl": true,
  "smtp_timeout": 15,
  "smtp_username": "**@********.**",
  "templates_folder": "data/templates",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": true,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

### Steps to reproduce
I Want to send on server a zip file with a custom duration of expiration (1 year)

### Expected behaviour
I wanted to pushed the file on server. 

### Actual behaviour
i got an error 502 Bad gateway on response. if i change the duration to 30 days the file upload well.

**Comment by BlackDex** [comment]
confirmed. Currently custom dates are not yet supported.

**Comment by cksapp** [comment]
I'm not sure how applicable this is, but I have not seen the same issue you experience while in the web vault.

Upon creating a send with a file attachment, the deletion date was set to 30 days _(can be set to a maximum of 31 days per [Bitwarden documentation](https://bitwarden.com/help/article/send-lifespan))_ and the expiration date was set over 1 year in advance for testing. The send was saved and uploaded, and the deletion dates and expirations dates were kept as expected.
![2021-10-06_11h10_51](https://user-images.githubusercontent.com/66083310/136232233-4450161c-70b0-4535-b818-35a34192195a.png)

Link for send is valid, and shows proper expiration date.
![2021-10-06_11h10_59](https://user-images.githubusercontent.com/66083310/136232235-169ec01c-680a-4882-a416-f5a7915e3cc8.png)

One thing I will note is that Bitwarden considers sends to be ephemeral and thus currently are limited only be set for a max of 31 days until deletion _(if trying to set expiration longer than 31 days an error occurs with message "You cannot have a Send with a deletion date that far into the future. Adjust the Deletion Date to a value less than 31 days from now and try again.")_, expiration is expected to be either not set, to where the send is deleted on deletion date and access to the send is removed once send has been deleted.
If needing expiration to be shorter than deletion, then the send can be set to expire **prior** to deletion date to remove access while still retaining the send before it becomes deleted automatically on the deletion date.

**Comment by BlackDex** [comment]
Ah, indeed, i haven't checked the bitwarden site it self yet, so didn't noticed they had a limit of 31 days.
So, in that case, i prefer to have it the same as upstream.
